### DISCUSSION TITLE
Hide the user agent in DoH requests to be more anonymous.

### DISCUSSION THREAD

**Discussion by PeterDaveHello** [DNSCrypt/dnscrypt-proxy#2013] [Ideas]
Recently, I found that `dnscrypt-proxy` is using the user agent `dnscrypt-proxy`, to those DNS-over-HTTPS servers, iOS and Firefox are both using empty user agent string.

Just wondering if you'd like to use an empty user agent string here to make the DNS requests more anonymous? Would like to send a pull request if it's acceptable, thanks!

**Comment by ghost** [comment]
Good idea. I guess if you only remove it for DoH and keep it for DNSCrypt-queries Frank might accepts it.

**Reply by PeterDaveHello** [reply]
It's a http request header, I'm not sure if it exists in DNSCrypt protocol.

**Reply by ghost** [reply]
Hmm, I remember someone asking to remove some client related information from DNSCrypt and Frank replied something like that wouldn't be necessary because at least all DNSCrypt users would look the same. Can't find it now but I don't think it was about DoH.

**Reply by jedisct1** [reply]
It can be removed, but I found it convenient for debugging. It doesn't leak any personal information either.

**Reply by jedisct1** [reply]
The DNSCrypt protocol doesn't allow including any extra information beyond the DNS query. Neither does Anonymized DNSCrypt.

**Reply by PeterDaveHello** [reply]
Yeah, it doesn't contain any personal info, just would be part of fingerprinting to profile a user and the behavior of the user. For the debugging purpose, maybe could have a debug mode? I guess at this moment, users using DoH, especially using dnscrypt-proxy, would like to keep privacy more :)

**Reply by PeterDaveHello** [reply]
@jedisct1 do you mind me sending a pull request to remove it?

**Reply by PeterDaveHello** [reply]
I just made a very tiny patch here for a little bit privacy enhancement: https://github.com/PeterDaveHello/dnscrypt-proxy/commit/8b141a9233b85704a6bfa566bc2f7a108fe15f0b

**Comment by PeterDaveHello** [comment]
Just another point I'm not sure, dnscrypt-proxy also sends query string `body_hash` in the requests, didn't find other DoH client's do the same yet, it'll be part of the anonymous if it's a unique behavior. @jedisct1 would you like to give us some comments?

**Reply by jedisct1** [reply]
`body_hash` is useful for caches and CDNs that don't use the body content in order to compute the cache key.

Without this, clients may end up receiving a cached response that doesn't match their query.

**Reply by PeterDaveHello** [reply]
That's interesting, but wouldn't Firefox, iOS, or other DoH face this problem? Maybe this is something that should be handled by the DoH providers. By default, I doubt that CDN would cache it.

According to [RFC 2616 Section 9.5, the POST method definition](https://datatracker.ietf.org/doc/html/rfc2616/#section-9.5):
> Responses to this method are not cacheable, unless the response includes appropriate Cache-Control or Expires header fields.

**Reply by PeterDaveHello** [reply]
Another problem to cache it is the DNS RR TTL issue, the cache should be done at the DNS level I guess, not the intermediate level like https server?

**Reply by PeterDaveHello** [reply]
BTW, Maybe using HTTP GET method would be easier in this case, because the data is in the URL.

**Reply by PeterDaveHello** [reply]
@jedisct1 not sure if you have any new thoughts that we can discuss together 😄

**Reply by jedisct1** [reply]
When using a CDN you want it to do caching to reduce latency and the load on your server. Or else the CDN is pointless 😂

I think dnscrypt-proxy tries to use GET by default but not all servers support it.

**Reply by PeterDaveHello** [reply]
Wouldn't CDN make the result not reliable as it has a different cache TTL?

**Reply by PeterDaveHello** [reply]
@jedisct1 for those servers doesn't support the GET method, does `body_hash` still make sense? Does `dnscrypt-proxy` try http GET first, and use POST for those who don't support http GET? (doesn't look like so)

**Reply by jedisct1** [reply]
Yes! `body_hash` is specifically useful for `POST`. Fastly caches resources based on the URL and some headers, not the `POST` content. It's possible to handle the content, but that requires custom code, is slower and more expensive. With default settings, things appear to work during testing, except that people may end up seeing other people's traffic. 

 Cloudflare doesn't allow caching `POST` requests at all except with Cloudflare Workers, hence `GET` being preferred. `GET` requests are also less likely to be compressed, which improves security.

**Reply by PeterDaveHello** [reply]
Sounds like Fastly implemented a different cache policy? Not sure there are too many CDN or proxy cache normal POST requests.

**Reply by jedisct1** [reply]
They have this notion of cache keys. A cache key is a hash of whatever you want from the `HEAD` portion of the request. Other CDNs probably work in a similar way. `POST` queries can be cached if the method is not part of the hash. But then you still need something to make the distinction between different queries.

I've no idea how to include the `POST` query content with `nginx` either. Having that hash solves the problem without having to worry about it.

`doh-server` adds HTTP caching headers based on the DNS records TTL, so caching at CDN level is fine, the original TTL is respected.

**Reply by PeterDaveHello** [reply]
Cool, I will try to figure it out, might not understand enough yet  😅
